### ISSUE TITLE
IDistributedCache based HubLifetimeManager implementation

### ISSUE THREAD

**Issue by laskoviymishka** [aspnet/SignalR#960]
Currently AspNetCore have `IDistributedCache` abstraction layer which is quite nice for having your own implementation of cache.

I believe current back plane with redis is better then nothing, but redis support for dotnet core is not perfect. StackExchange.Redis has a some major issues with performance and right now completely useless in docker based enviroment. It not even support sentinel and ipv6 for now. So i belive stick with redis only is not good option.

May be generic back plan with `IDistributedCache` solve issue of redis support?

Some people may event has own implementation of `IDistributedCache` interface that suite for them business (me for example :shipit: ).

I not saying to remove `Signalr.Redis`, i talk about something more generic like `Signalr.DistributedCache`.

**Comment by davidfowl** [comment]
I don't see how this makes any sense as the backplane implementation is much more than a cache. Using the IDistributedCache you wouldn't get access to redis' pubsub dewryres which is the only thing we use in the backplane implementation.

Can you clarify what you think an implementation would look like?

**Comment by laskoviymishka** [comment]
May be `IDistributedCache` is not a best option, but definetely having only redis implementation made SignalR not usefull.

Right now Redis is unusable for some scenarios (see https://github.com/StackExchange/StackExchange.Redis/issues/410 and https://github.com/StackExchange/StackExchange.Redis/issues/252). 

For example you have an Application that hosted somewhere in a cloud via docker, and it has 3 instances. I wanna use websocket inside of this app to notify some user groups. And right now (without any other backplane rather then redis) i can't use it at all. As far as i understand the main problem that need to be solved - sharing `HubConnectionList` between instances. It may be implemented on top of `IDistributedCache` - if we store information about wich instances has which connection id's and delegete notification of a right users to a right instance.

Let continue 3 (A,B,C) instances example. We have 3 servers with 4 connection that lives in 3 groups with a follow distribution:

| connection   	| instance  	| group  	|
|---	|---	|---	|
| con1	        |A	                | gr1,  gr3	|
| con2   	        |B   	                | gr1, gr2, |
| con3   	        |C   	                |  gr2, gr3 |
| con4   	        |B   	                | gr1,        |

Our distributed cache may store something like that.
Let me made some assumptions:

- Let assume that each server connected with each other via SignalR server -> server connection or may be connected any time.

- Let assume that we have an HTTP api that simply broadcast text message to some group.

- Let assume that our application hide beside some load balancer that randomly navigate user to any of 3 instances

So when i make a call to this HTTP Api to notify gr1, i should notify 3 users with con1, 2 and 4 with message 'hello'. 
Let's imagine that this call arrived at instance C, and c# code call `InvokeGroupAsync` of this Imagine `DistrCacheLifetimeManager`. Inside of it simply look into distr cache for all connection and filter desired group, than batch them over instance and send some internal message to instance A to notify con1, instance B to notify con2 and con4 with message 'hello'.

It not so straight forward, but not so far from current Redis implementation.

-----

Bottom line is - we need some other option for backplane (may be some mq like rabbit?)

**Comment by davidfowl** [comment]
Thanks for the feedback, we have plans to ship redis and service bus implementations (and possibly sql server). You can also write your own backplane (as many others have done in the past e.g. https://github.com/search?utf8=%E2%9C%93&q=signalr+backplane&type=)

**Comment by davidfowl** [comment]
Feel free to prototype the approach mentioned. It would give great insight on how to potentially come up with a solution that does what you want. I'd also recommend looking at clustering solutions to solve some of the communication problem between nodes

**Comment by davidfowl** [comment]
I filed this to investigate something https://github.com/aspnet/SignalR/issues/968
